domain/application/state/applicationresource.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/collections/set"
+
+	coreresource "github.com/juju/juju/core/resource"
+	"github.com/juju/juju/domain/application"
+	"github.com/juju/juju/domain/application/charm"
+	apperrors "github.com/juju/juju/domain/application/errors"
+	"github.com/juju/juju/internal/errors"
+)
+
+// buildResourceToAdd creates resources to add based on provided app and charm
+// resources.
+// Returns a slice of resourceToAdd and an error if any issues occur during
+// creation. Validates app resources against charm resources, warning if
+// mismatches are found.
+// Returns an error satisfying [apperrors.InvalidResourceArgs] if there is
+// unexpected resource in argument that are not defined in charm.
+func (st *State) buildResourceToAdd(appUUID, charmUUID string,
+	appResources []application.AddApplicationResourceArg,
+	charmResources map[string]charm.Resource) (
+	[]resourceToAdd, error) {
+	resources := make([]resourceToAdd, 0, len(appResources))
+	appResourceSet := set.NewStrings()
+	charmResourceSet := set.NewStrings()
+	for _, r := range appResources {
+		appResourceSet.Add(r.Name)
+		uuid, err := coreresource.NewUUID()
+		if err != nil {
+			return nil, errors.Capture(err)
+		}
+		resources = append(resources, resourceToAdd{
+			UUID:            uuid.String(),
+			ApplicationUUID: appUUID,
+			CharmUUID:       charmUUID,
+			Name:            r.Name,
+			Revision:        r.Revision,
+			Origin:          r.Origin.String(),
+			State:           "available", //todo(gfouillet): use state.go once #18460 landed
+			CreatedAt:       st.clock.Now(),
+		})
+	}
+	for _, res := range charmResources {
+		charmResourceSet.Add(res.Name)
+	}
+
+	unexpectedResources := appResourceSet.Difference(charmResourceSet)
+	missingResources := charmResourceSet.Difference(appResourceSet)
```

Comment:
This can be done in the service.

Reply:
I agree it can. But from my perspective it shouldn't.

I think that it is the responsibility of the state layer to ensure that the state is consistent, and provide useful error or log whenever it may be compromised.

Here, not having a 1-1 match between charm resources and inserted resources causes an inconsistent state, and in some case (more inserted resources than existing in the charm), would cause a constraint error, with few details (at least I get few details in the TU), which will not be easy to deal with when it will return as an issue.

Checking it in the state before insertion, coupled with TU, is a good thing for this reason, and we could also check it in the service but it would be overkill.

Service layer only check that resources args are consistent (valid, with origin consistent with possible revision) one by one, but consistency check is better done in state, IMHO.

This way, if we choose to move the functionality to solve resources in the service for some reason (instead of in the caller), we cannot break the consistency of the state by mistake.